src/devices/machine/ibmps2.h

```diff
@@ -0,0 +1,265 @@
+	auto kbd_data() { return subdevice<ps2_keyboard_controller_device>("keybc")->kbd_data(); }
+
+	DECLARE_WRITE_LINE_MEMBER(kbd_clk_w);
+	DECLARE_WRITE_LINE_MEMBER(kbd_data_w);
+
+	void write_rtc(offs_t offset, uint8_t data);
+	
+	DECLARE_WRITE_LINE_MEMBER(iochck_w);
+	DECLARE_WRITE_LINE_MEMBER(shutdown);
+
+	void at_softlists(machine_config &config);
+
+    void    post_debug_w(offs_t offset, uint8_t data);
+
+    uint8_t memory_control_r();
+    void    memory_control_w(uint8_t data);
+
+    uint8_t nvram_r(offs_t offset);
+    void    nvram_w(offs_t offset, uint8_t data);
+
+	template<typename R, typename W> void install_memory(offs_t start, offs_t end, R rhandler, W whandler)
+	{
+		m_mcabus->install_memory(start, end, rhandler, whandler);
+	}
+
+	IRQ_CALLBACK_MEMBER(irq_ack_cb);
+
+protected:
+	required_device<cpu_device> m_maincpu;
+    required_device<ram_device> m_ram;
+	optional_device<nvram_device> m_nvram;
+	required_device<mca16_device> m_mcabus;
+	required_device<pic8259_device> m_pic8259_master;
+	required_device<pic8259_device> m_pic8259_slave;
+	required_device<ibm72x7377_device> m_dmac;
+	required_device<ps2_pit_device> m_pit;
+	required_device<speaker_sound_device> m_speaker;
+	required_device<mc146818_device> m_mc146818;
+	required_device<ps2_keyboard_controller_device> m_keybc;
+    required_device<pc_kbdc_device> m_ps2_con;
+	required_device<pc_kbdc_device> m_aux_con;
+	required_device<mca16_slot_device> m_planar_fdc;
+	required_device<mca16_slot_device> m_planar_uart;
+	required_device<mca16_slot_device> m_planar_lpt;
+	required_device<mca16_slot_device> m_planar_vga;
+
+	optional_device<mca16_slot_device> m_mcaslot_1;
+	optional_device<mca16_slot_device> m_mcaslot_2;
+	optional_device<mca16_slot_device> m_mcaslot_3;
+	optional_device<mca16_slot_device> m_mcaslot_4;
+	optional_device<mca16_slot_device> m_mcaslot_5;
+	optional_device<mca16_slot_device> m_mcaslot_6;
+	optional_device<mca16_slot_device> m_mcaslot_7;
+	optional_device<mca16_slot_device> m_mcaslot_8;
+
+	uint16_t m_sram_size;
+	uint16_t m_planar_id;
+
+	std::unique_ptr<uint8_t[]> m_sram;
+    uint16_t m_nvram_index = 0;
+
+	void ps2_32_map(address_map& map);
+	void ps2_32_io(address_map& map);
+
+	void ps2_16_map(address_map& map);
+	void ps2_16_io(address_map& map);
+
+	void device_start() override;
+	void device_reset() override;
+	virtual void device_add_mconfig(machine_config &config) override;
+    virtual void device_config_complete() override;
+
+	void add_mca16(machine_config &config);
+	void add_mca32(machine_config &config);
+
+	void speaker_set_spkrdata(uint8_t data);
+
+	uint8_t m_at_spkrdata = 0;
+	uint8_t m_pit_out2 = 0;
+	bool m_write_gate_2 = 0;
+	
+	uint8_t m_at_speaker = 0;
+	uint8_t m_channel_check = 0;
+	uint8_t m_nmi_enabled = 0;
+	uint8_t m_parity_check_enabled = 0;
+	uint8_t m_channel_check_enabled = 0;
+	uint8_t m_memory_control = 0;
+	bool m_gate_a20 = 0;
+	bool m_nmi_flag = 0;
+	bool m_alt_hot_reset = 0;
```

Comment:
Please don’t use inline initialisers in device headers.  It’s makes things more fragile and adds potential for confusion.  Also this file uses a mix of tabs and spaces for indentation.